### DISCUSSION TITLE
Is it possible to share the same `#[pyclass]` instance as argument to python function in rust？

### DISCUSSION THREAD

**Discussion by sun-rs** [PyO3/pyo3#3010] [Questions] [answered]
Assume created rust and python code like below:

```python
class UserDef:
    def on_event(self, ctx:Context):
        ctx.some_data += 9

u = UserDef()
c = Context(u)

Executor().run_backtest(c)
```

```rust
#[pyclass]
struct Context {
    pub some_data: i32
    pub uesrdef: PyAny
}

#[pymethod]
impl Context {
    #[new]
    pub new(userdef: PyAny) -> Self {
        Self {
            some_data: 1,
            userdef
        }
    }

    pub fn add(&mut self) {
        self.some_data += 111;
    }

#[pyclass]
struct Executor {}

#[pymethod]
impl Executor {
    pub fn run_backtest(&self, ctx: &mut Context) {
        for i in 0..10000 {
            ctx.add()
            ctx.userdef.call_method1("on_event", ctx);

        }
    }

```
I want to share the same `Context` instance between python and rust as a context manager to deal with data. Rust side would change its data no matter what, also I want give python side some freedom to change ctx data as the user wish, the `on_event` is flexible.  Could this be possible without copy data？

**Comment by adamreichold** [comment, marked as answer]
`on_event` will want access to something like `&PyAny` as it will not be able to handle the Rust type `Context`. So maybe something like

```rust
    pub fn run_backtest(&self, ctx: &PyCell<Context>) {
        for i in 0..10000 {
            ctx.borrow_mut().add();
            ctx.userdef.call_method1(intern!(ctx.py(), "on_event"), ctx);
        }
    }
```

?

**Reply by sun-rs** [reply]
Thanks for reply!  Could I consider this whole process deal with a same instance of Context without copy from first to last？

And could you please give some advice about multiple #[pyclass]  ´Context´ in python's multiprocess read(almost no write) same shared big data(with different length of slice on this data)？

**Reply by adamreichold** [reply]
> Could I consider this whole process deal with a same instance of Context without copy from first to last？

Yes.
